### PR TITLE
Health checks on Ray Deployments

### PR DESCRIPTION
**Pull request by korinayo**
Adds basic health check functionality that runs every 30s. Callable via inference API. 

Tested for all deployment types we have in prod.

### CODE REVIEW

File: llama_vision.py [hunk 1]
```diff
@@ -194,6 +194,41 @@ def model_config(self):
     def model_device(self) -> str:
         return str(self.model.device)
 
+    @web_app.get("/health")
+    def check_health(self):
+        """Health check that verifies basic model functionality."""
+        try:
+            self._clear_cache()
```
Comment (dienhartd) [inline llama_vision.py:201]: sanity check: does `_clear_cache()` every 30s affect anything else as a side effect or just this route?

**Review by dienhartd — COMMENTED** [review]
one sanity check question otherwise lgtm

File: llama_vision.py (hunk 1 above)
Comment (korinayo) [reply llama_vision.py:201]: affects the model object hosted at this deployment as well.

File: mini_cpm.py [hunk 2]
```diff
@@ -194,6 +194,31 @@ async def image_infer(
         finally:
             self._clear_cache()
 
+    @web_app.get("/health")
+    def check_health(self):
```
Comment (boyleconnor) [inline mini_cpm.py:198]: Do I understand that Ray guarantees that `deployment`s will have their `check_health()` method called automatically (on a user-defined period)? The [documentation](https://docs.ray.io/en/latest/serve/api/doc/ray.serve.deployment_decorator.html) seems to imply as much, but in a very off-handed sort of way, without much detail.

File: transformers_model.py [hunk 3]
```diff
@@ -163,6 +163,32 @@ def infer(self, inference_request: InferenceRequest) -> dict:
...
+        """Health check that verifies basic model functionality."""
+        try:
+            self._clear_cache()
+
+            # Basic inference test
+            with torch.no_grad():
+                if self.task == "text-generation":
+                    self.pipe("Is this thing on?", max_new_tokens=10)
+                else:
+                    self.pipe("Is this thing on?")
```
Comment (boyleconnor) [inline transformers_model.py:177]: I'm not confident that these two formats cover all potential `task` schemas. Are you, @korinayo?

**Comment by boyleconnor** [comment]
Have you tested this? One way to do so would be to create a branch in `elevaite` (branched off `main`) with the [`DEFAULT_WORKING_DIR`](https://github.com/bracket-ai/elevaite/blob/a9ad6e39bd6b9fa31dcf293660d7ad5fc201801b/elevaite/rayservice_utils.py#L30) named constant changed to this branch (or even better, a specific commit thereof). You could also change the test logic to add a call to `/health` (e.g. around [here](https://github.com/bracket-ai/elevaite/blob/a9ad6e39bd6b9fa31dcf293660d7ad5fc201801b/tests/e2e/test_web_server.py#L97))

File: mini_cpm.py (hunk 2 above)
Comment (korinayo) [reply mini_cpm.py:198]: Yep -- called automatically. 10s intervals is default.

File: transformers_model.py (hunk 3 above)
Comment (korinayo) [reply transformers_model.py]: Not confident -- was biasing too heavily to current use cases. Updated

**Comment by korinayo** [comment]
> Have you tested this? One way to do so would be to create a branch in `elevaite` (branched off `main`) with the [`DEFAULT_WORKING_DIR`](https://github.com/bracket-ai/elevaite/blob/a9ad6e39bd6b9fa31dcf293660d7ad5fc201801b/elevaite/rayservice_utils.py#L30) named constant changed to this branch (or even better, a specific commit thereof). You could also change the test logic to add a call to `/health` (e.g. around [here](https://github.com/bracket-ai/elevaite/blob/a9ad6e39bd6b9fa31dcf293660d7ad5fc201801b/tests/e2e/test_web_server.py#L97))

have tested on the on prem test cluster @boyleconnor

File: transformers_model.py [hunk 4]
```diff
@@ -163,6 +163,33 @@ def infer(self, inference_request: InferenceRequest) -> dict:
...
+
+        try:
+            self._clear_cache()
+
+            # Basic inference test
+            with torch.no_grad():
+                self.pipe("Is this thing on?", max_new_tokens=10)
+
+            logger.info("Health check passed")
+            return {"status": "healthy"}
```
Comment (boyleconnor) [inline transformers_model.py:181]: In the case where `task != "text-generation"`, can you change `"status"` to `"unknown"` (I'm assuming Ray will just check the status code/whether the method completes without raising an exception).

If that doesn't work (i.e. a different value of `status` causes the health check result to be misinterpreted as unhealthy), can you try adding another key to this dictionary such as `"info"` with a message such as `f"warning: task '{self.task}' does not currently have a working health check"`.

Basically, I'd like there to be some indication that this health check is not doing what it promises to do.

File: transformers_model.py [hunk 5]
```diff
@@ -163,6 +163,33 @@ def infer(self, inference_request: InferenceRequest) -> dict:
...
+
+        except Exception as e:
+            self._clear_cache()
+            logger.error(f"Health check failed: {e}", exc_info=True)
+            raise HTTPException(
+                status_code=HTTPStatus.INTERNAL_SERVER_ERROR,
+                detail="Endpoint is unhealthy. Basic model.pipe() call failed.",
```
Comment (boyleconnor) [inline transformers_model.py:188]: Nitpick: the call isn't to `self.model.pipe()`, it is to `self.pipe()`. This should say something like `Basic inference call failed.`

File: transformers_model.py [hunk 6]
```diff
@@ -163,6 +163,33 @@ def infer(self, inference_request: InferenceRequest) -> dict:
     def model_config(self):
         return numpy_to_std(self.pipe.model.config.__dict__)
 
+    @web_app.get("/health")
+    def check_health(self):
+        # We only run this for text-generation tasks at the moment
+        # TODO: Add support for other tasks
+        if self.task != "text-generation":
+            return
```
Comment (boyleconnor) [inline transformers_model.py:171]: Putting this outside the `try...except...finally` block means that the cache won't be cleared during health check calls; is this the desired behavior?

Comment (boyleconnor) [reply transformers_model.py:171]: Actually, I can't tell if this is intended to return "healthy" by default for non-text-generation endpoints, or "unhealthy"

File: transformers_model.py (hunk 4 above)
Comment (boyleconnor) [reply transformers_model.py:181]: ^this comment is premised on the assumption that we return "healthy" by default for all non-text-generation tasks. Ignore if that is not true

**Review by boyleconnor — CHANGES_REQUESTED** [review]

File: transformers_model.py (hunk 6 above)
Comment (korinayo) [reply transformers_model.py:171]: @boyleconnor  It returns nothing! The health check serves a dual purpose here. 

- For Ray: any code that returns an error message will cause a failed Ray health check that will change the status of the deployment. Returning nothing will trigger neither pass nor failure. 
- For elevaite: Binu requested something that returns a health status for the endpoint if we know it. I think your request for an "unknown" status is correct for this purpose. As is the response body would be empty. If we can't replicate the call structure of the model then the endpoint could very well be healthy and we wouldn't know.

File: transformers_model.py (hunk 4 above)
Comment (korinayo) [reply transformers_model.py:181]: the assumption is incorrect, but the warning message is a good idea.

File: transformers_model.py (hunk 6 above)
Comment (boyleconnor) [reply transformers_model.py:171]: Ah, thanks for the clarification. Thiis is resolved, assuming [this warning message](https://github.com/bracket-ai/elevaite-inference-rayservice/pull/21#discussion_r1893495370) gets implemented

File: transformers_model.py (hunk 3 above)
Comment (korinayo) [reply transformers_model.py]: further update here: have manually tested all of elevaite's [supported tasks](https://github.com/bracket-ai/elevaite/blob/main/elevaite/models.py#L8-L23) and come up with a set that work out of the box with this current pipe(text) call. We will run this version of the health check for all deployments w/ pipelines of these tasks. I will create an issue to figure out how to expand others (if feasible). For now this covers pretty much every current elevaite use case we've

**Review by boyleconnor — APPROVED** [review]

**Comment by boyleconnor** [comment]
@korinayo hold on, I think you should bump the version number in `pyproject.toml`

**Comment by korinayo** [comment]
> @korinayo hold on, I think you should bump the version number in `pyproject.toml`

thank you! done!

File: transformers_model.py [hunk 7]
```diff
@@ -163,6 +174,39 @@ def infer(self, inference_request: InferenceRequest) -> dict:
...
+            # we don't currently support health checking them
+            # FIXME: Add support for health checking these tasks.
+            # Will require matching the more complex call signature of these tasks.
+            if self.task not in SUPPORTED_HEALTH_CHECK_TASKS:
+                return {"warning": f"Health check not supported for task {self.task}"}
+
+            # Basic inference test
+            # If this errors, the health check will fail.
+            with torch.no_grad():
+                self.pipe("Is this thing on?", max_new_tokens=10)
```
Comment (boyleconnor) [inline transformers_model.py:195]: This `max_new_tokens` should only be there for `text-generation` models. Fixing this in #22
